File: back/src/main/java/com/back/domain/schedule/entity/Schedule.java
package com.back.domain.schedule.entity;

import com.back.domain.concert.entity.Concert;
import com.back.global.jpa.entity.BaseEntity;
import com.back.domain.venue.entity.Venue;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Schedule extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long scheduleId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "concert_id", nullable = false)
    private Concert concert;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "venue_id", nullable = false)
    private Venue venue;

    @Column(nullable = false)
    private LocalDateTime scheduleDate;

    @Column(nullable = false)
    private int round;

    private Schedule(Concert concert, Venue venue, LocalDateTime scheduleDate, int round) {
        this.concert = concert;
        this.venue = venue;
        this.scheduleDate = scheduleDate;
        this.round = round;
    }

    public static Schedule create(Concert concert, Venue venue, LocalDateTime scheduleDate, int round) {
        return new Schedule(concert, venue, scheduleDate, round);
    }
}

File: back/src/main/kotlin/com/back/domain/schedule/entity/Schedule.kt
package com.back.domain.schedule.entity

import com.back.domain.concert.entity.Concert
import com.back.domain.venue.entity.Venue
import com.back.global.jpa.entity.BaseEntity
import jakarta.persistence.*
import java.time.LocalDateTime

@Entity
open class Schedule protected constructor() : BaseEntity() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val scheduleId: Long? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "concert_id", nullable = false)
    lateinit var concert: Concert
        protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "venue_id", nullable = false)
    lateinit var venue: Venue
        protected set

    @Column(nullable = false)
    lateinit var scheduleDate: LocalDateTime
        protected set

    @Column(nullable = false)
    var round: Int = 0
        protected set

    private constructor(
        concert: Concert,
        venue: Venue,
        scheduleDate: LocalDateTime,
        round: Int
    ) : this() {
        this.concert = concert
        this.venue = venue
        this.scheduleDate = scheduleDate
        this.round = round
    }

    fun isExpired(): Boolean = LocalDateTime.now().isAfter(
        scheduleDate.toLocalDate().minusDays(1).atTime(17, 0)
    )

    companion object {
        @JvmStatic
        fun create(concert: Concert, venue: Venue, scheduleDate: LocalDateTime, round: Int): Schedule =
            Schedule(concert, venue, scheduleDate, round)
    }
}
